crates/bevy_app/src/app.rs

```diff
@@ -56,6 +56,9 @@ pub struct App {
     /// A container of [`Stage`]s set to be run in a linear order.
     pub schedule: Schedule,
     sub_apps: HashMap<Box<dyn AppLabel>, SubApp>,
+    /// A private marker to prevent incorrect calls to `App::run()` from `Plugin::build()`
+    /// <https://github.com/bevyengine/bevy/issues/4231>
+    is_from_plugin_build: bool,
```

Comment:
Super nit: I think calling this something like `is_building_plugin` is a little easier to understand.